.github/workflows/codecov_ats.yml

```diff
@@ -80,13 +87,43 @@ jobs:
           env:
             CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
             CODECOV_STATIC_TOKEN: ${{ secrets.CODECOV_STATIC_TOKEN }}
-        # Run label analysis IN DRY MODE (no tests will actually run)
-        - name: Label Analysis
+        # Run Automated Test Selection in dry mode to get the list of tests to run
+        # The base commit will be the parent commit (apparently commits on master don't exist in codecov)
+        - name: Codecov Automated Test Selection
+          id: codecov_automated_test_selection
           continue-on-error: true
           run: |
-            BASE_COMMIT=$(git merge-base ${{ github.sha }}^ origin/master)
+            BASE_COMMIT=$(git rev-parse ${{ github.sha }}^)
             echo $BASE_COMMIT
-            codecovcli --codecov-yml-path=codecov.yml label-analysis --dry-run --token=${CODECOV_STATIC_TOKEN} --base-sha=${BASE_COMMIT}
+            output=$(codecovcli --codecov-yml-path=codecov.yml label-analysis --dry-run --token=${CODECOV_STATIC_TOKEN} --base-sha=${BASE_COMMIT}) || true
+            if [ -n "${output}" ];
+            then
+              json_output=$(echo $output | grep -o -e "{.*}")
+
+              echo ATS_TESTS_TO_RUN=$(jq <<< $json_output '.runner_options + .ats_tests_to_run | map(@sh) | join(" ")' --raw-output) >> "$GITHUB_OUTPUT"
+              echo ATS_TESTS_TO_SKIP=$(jq <<< $json_output '.runner_options + .ats_tests_to_skip | map(@sh) | join(" ")' --raw-output) >> "$GITHUB_OUTPUT"
+
+              test_to_run_count=$(echo $json_output | jq '.ats_tests_to_run | length')
+              test_to_skip_count=$(echo $json_output | jq '.ats_tests_to_skip | length')
+              total_tests_count=$(($test_to_run_count + $test_to_skip_count))
+              echo "Selected $test_to_run_count / $total_tests_count tests to run" >> $GITHUB_STEP_SUMMARY
+            else
+              tee <<< "ATS failed. Can't get list of tests to run. Fallback to all tests" "$GITHUB_STEP_SUMMARY"
+              echo "ATS_TESTS_TO_RUN=--cov-context=test" >> "$GITHUB_OUTPUT"
+              echo "ATS_TESTS_TO_SKIP=" >> "$GITHUB_OUTPUT"
+            fi
           env:
             CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
             CODECOV_STATIC_TOKEN: ${{ secrets.CODECOV_STATIC_TOKEN }}
+    # The actual running of tests would come here, after the labels are available
+    # Something like pytest <options> $ATS_TESTS_TO_RUN
+    debug:
+      runs-on: ubuntu-latest
+      needs: coverage-ats
+      steps:
+        - name: Debug ATS_TESTS_TO_RUN
+          run: |
+            echo ${{ needs.coverage-ats.outputs.ATS_TESTS_TO_RUN }}
```

Comment:
"echo" doesn't give an accurate picture of quoting:

```console
$ echo a "b c"
a b c
```

The best thing to do for logging is `set -x`:

```console
$ set -x

$ : a "b c"
+ : a 'b c'
```

I actually recommend adjusting the default shell options to make this default:

```yaml
defaults:
  run:
    # the default default is:
    #      bash --noprofile --norc -eo pipefail {0}
    shell: bash --noprofile --norc -eo pipefail -ux {0}
```